text/0161-telemetry-schema-and-resource.md

```diff
@@ -0,0 +1,203 @@
+# Telemetry Schema and Resource
+
+Specify a mechanism for users of Resource to interact with Resource leveraging
+TelemetrySchema version for a stable API.
+
+## Motivation
+
+In the [current definition](https://github.com/open-telemetry/opentelemetry-specification/pull/1692)
+of Resource + Telemetry schema interaction, the specification focuses mostly on
+how to provide a consistent "storage" of Resource within the SDK. However,
+from a usage standpoint, how do consumers gain stability across changes to
+schema?
+
+- Telemetry Schema allows the notion of "renaming" attributes in a schema.
+  Can a consumer of Resource "lock" in a specific version of the schema to
+  ensure no renaming occurs?
+- If resource detection is provided by libraries external to open-telemetry, do
+  we require ALL resource detectors to be against the same telemetry schema
+  version number?  This would inhibit upgrading versions as you'd need to wait
+  for all resource detectors to upgrade, even if a (stable) conversion exists
+  from the resource version.
+
+Specifically, we need a way for exporters that rely on Resource labels to ensure
+various version differences across OpenTelemetry components do not cause
+breakages by leveraging the powers provided by telemetry schema.
+
+As an example, prior to schema version and stability, the PR
+[#1495](https://github.com/open-telemetry/opentelemetry-specification/pull/1495)
+renamed `cloud.zone` to `cloud.availability_zone`.  In a Schema world, this
+change could have been defined with the following schema rule:
+
+```yaml
+  next.version:
+    resources:
+      changes:
+        - rename_attributes:
+            attribute_map:
+              cloud.zone: cloud.availability_zone
+  previous.version:
+```
+
+If an exporter is defined expecting to find `cloud.zone` on `previous_version`,
+then when the SDK is bumped to `next.version` it will no longer be able to see
+`cloud.zone`.
+
+Specifically, we assume the following components could have independent
+versioning + release cycles:
+
+```
++-------------------+     +-----+     +----------+
+| Resource Detector | <-> | SDK | <-> | Exporter |
++-------------------+     +-----+     +----------+
+```
+
+## Explanation
+
+We propose a new set of SDK features:
+
+- A mechanism to "migrate" a resource up/down compatible schema changes.
```

Comment:
I think we should provide only the "go up" version, and enforce everyone to use the most recent (latest) schema. The exporter in your case should be able to adopt (maybe) the rules to work on the Resource schema (if newer) vs downgrading the Resource.

Reply:
We should be comfortable with both up and down directions. The schema files unambiguously defines whether it is safe to go in a particular direction.